### ISSUE TITLE
Support PassKey Integration (Web Authentication API)

### ISSUE THREAD

**Issue by Ajedi32** [keepassxreboot/keepassxc#1870]
Recently, the W3C finalized their Candidate Recommendation for the [Web Authentication Standard](https://www.w3.org/TR/webauthn/), a new browser feature that allows sites to request user authentication through a standardized API.

Implementation is currently underway in [Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=664630), [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1384776), and [Edge](https://developer.microsoft.com/en-us/microsoft-edge/platform/status/webauthenticationapi/?q=web%20authentication).

Once this standard starts being implemented by sites, I'd like to be able to use KeePassXC as an [Authenticator](https://www.w3.org/TR/webauthn/#sctn-authenticator-model) for it, so that I can store my Web Authentication private keys in the same place my existing legacy passwords are stored, and keep them synchronized across all my devices.

While it might still be too early to begin working on a concrete implementation (available documentation for the Web Authentication API is still a little sparse, outside of the standards document itself), I figured it can't hurt to start looking into this now so we're prepared when the time comes.

Would integration with the Web Authentication API be a feature worth supporting?

**Comment by varjolintu** [comment]
I will definitely keep an eye on this!

**Comment by Typz** [comment]
Web Authentication is supported in the latest Google Chrome release (67)

**Comment by varjolintu** [comment]
So far I haven't found a way to use a software as an Authenticator, and I assume it will not be possible unless you have a idiot-proof way to emulate a USB device.

**Comment by Ajedi32** [comment]
It's possible official support for third-party software authenticators could be coming in the future (though I've heard no news about anything like that yet); but even in the absence of that couldn't you use a browser extension to directly intercept calls to the Web Authentication API and implement it that way?

Until we have some real-world examples of sites using the Web Authentication API as a password replacement, it's hard to say what the best way to approach the problem is.

**Comment by Ajedi32** [comment]
FYI, the Web Authentication standard is now officially a W3C recommendation: https://www.w3.org/TR/2019/REC-webauthn-1-20190304/

It's possible to implement an Authenticator via a browser extension by intercepting calls to the Web Authentication API and returning an appropriate response to the web page. Krypton is a good example of how this can be done: https://github.com/kryptco/kr-u2f/blob/192af059419501bb702148ae5301f042c9859447/src/inject_webauthn_chromium.ts (Note the linked code is not FOSS so we can't copy it directly, but it does provide a general idea of how this can be done.)

WebAuthn can be tested here: https://webauthn.org/

**Comment by droidmonkey** [comment]
Cool, thank you for keeping on top of this

**Comment by rugk** [comment]
But is not WebAuthn intended to be used as a 2FA method? Not a replacement for passwords…

And if you use WebAuthn as a password replacement, you cannot really use any 2FA anymore? Or what should one do, if one then wants to use a hardware key as 2FA WebAuthn e.g.?

**Comment by droidmonkey** [comment]
WebAuthn is replacing human generated passwords with challenge response keys. These *CAN* be stored on hardware tokens, but certainly not necessary or desired in most cases. I for one hate carrying a token around with me.

**Comment by rugk** [comment]
>. I for one hate carrying a token around with me.

But you notice this is a security advantage?

So let's think again about 2FA: You want:
* something you know
* something you have
* something you are

So "traditionally" passwords are the first (you know them) and things like a TOTP app (on a different mobile phone) or hardware tokens are "what you have".
With KeePassXC, it would only be "what you know"…

Respectively it would all be on one device (being one factor in the end). If that is compromised, all your logins are compromised. Without it, you have currently have to compromise two devices (KeePass DB and on one phone or even a hardware key), which is obviously more secure…

**Comment by droidmonkey** [comment]
Security is all about _choice_. I did not say this is replacing 2FA or that it is stronger than 2FA. The vast majority of the world has no idea what 2FA is, nor do they care. I have lowered my personal risk on those accounts that I have deemed are "high value" by using true 2FA. For accounts that are not high value, I would very much like to have the convenience of using WebAuth without a hardware token.

You can make your own choices. That's the beautiful thing about KeePassXC!

**Comment by nagromc** [comment]
> 
> 
> > . I for one hate carrying a token around with me.
> 
> But you notice this is a security advantage?
> 
> So let's think again about 2FA: You want:
> 
>     * something you know
> 
>     * something you have
> 
>     * something you are
> 
> 
> (all and/or connected, 2 factors for two-factor-authenticatiion obviously)
> 
> So "traditionally" passwords are the first (you know them) and things like a TOTP app (on a different mobile phone) or hardware tokens are "what you have".
> With KeePassXC, it would only be "what you know"…
> 
> Respectively it would all be on one device (being one factor in the end). If that is compromised, all your logins are compromised. Without it, you have currently have to compromise two devices (KeePass DB and on one phone or even a hardware key), which is obviously more secure…

It is understandable and I agree with you. But as @droidmonkey said, it's a trade-off between security vs. convenience. And it is perfectly assumed in the [FAQ](https://keepassxc.org/docs/#faq-security-totp):

> We believe that **storing both together [passwords and TOTP secrets] can still be more secure than not using 2FA at all**, but to maximize the security gain from using 2FA, you should always store TOTP secrets in a separate database, secured with a different password, possibly even on a different computer.

You have the choice to store your 2FA (TOTP or FIDO2/CTAP) secrets in the same database.

**Comment by phoerious** [comment]
Nothing says you could not generate a WebAuthn response from more than one factor (I haven't read the spec, but it's quite flexible and generic from what I've heard).
As used today, 2FA is largely a mitigation of weak passwords and not of break ins into your machine. If your PC is taken over, then taking over your phone as well is usually rather easy (or vice versa). You only get the full advantage of 2FA if at least one of your tokens never comes in contact with the other (like, e.g. a YubiKey or a dedicated hardware token generator that is hard to impossible to penetrate with malware).

**Comment by rugk** [comment]
>  If your PC is taken over, then taking over your phone as well is usually rather easy (or vice versa).

I would totally challenge that assumption…

----

But yes, it can be totally useful as an optional feature. I totally think users can trade-off things there. (We have the same thing with TOTP tokens in KeePassXC.)

What I am just not sure about is, whether the WebAuthn spec actually intends this implementation? Does it take the thing into account (for security implications or so) or is the spec rather intended for 2FA?

**Comment by AWalton** [comment]
It's up to the website to support 2FA, I don't see any reason why WebAuthn can't be used as one factor. Similarly to how Google implements 2FA login it would just be WebAuthn on one page then TOTP or something else on the next.

**Comment by varjolintu** [comment]
I already have a test branch that can intercept WebAuthn create/get on the test page. For now it doesn't do anything yet, but I'll try to improve things when I have some time. But it's certainly possible to do.

**Comment by darkdragon-001** [comment]
What's the state of this?

**Comment by varjolintu** [comment]
@darkdragon-001 Haven't had time to do it any further. Not at simple as it sounds.

**Comment by nm17** [comment]
Any news on this?

**Comment by varjolintu** [comment]
@nm17 Yes. Sort of. I've tested some possibilities with this via the browser extension, but so far there are some security issues how to transfer certain messages from/to content scripts. At this point I haven't found a way to securely transfer some variables related to authentication without the possibility of other scripts capturing them as well.

**Comment by rugk** [comment]
Would it make sense to elaborate on that i.e./respectively document that progress somewhere (so other scan help/pick it up), or ask some question on Stackoverflow or similar about the problem you are facing?

**Comment by varjolintu** [comment]
@rugk Sure, I can provide some details. If anyone have better ideas for the implementation, please let me know.

To get the Webauthn to work, browser extension has to add a separate script file to `web_accesible_resources` in `manifest.json`, and inject that script to _every_ page. Yes. Without that there's no way you can override `navigator.credentials` with your own implementation (see details: https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API).
Injecting a script to every web page already gives me some shivers, but that's not all. If I want to send a message to the actual content script (and to KeePassXC from there) I'll have to use `window.postMessage()` and  `window.addEventListener('message', ...)` for the response, just because the script is injected. Transferring messages directly to the background script seems impossible, and using `window` for messages means every other script/extension can also listen to them. Much secure.

**Comment by Ajedi32** [comment]
It sounds like what you're looking for is [`browser.runtime.sendMessage`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendMessage). The KeePassXC Browser extension [already has a wrapper for that](https://github.com/keepassxreboot/keepassxc-browser/blob/a037a0b656201fdf9a467c243fd2b6c6a3f9d8b9/keepassxc-browser/content/keepassxc-browser.js#L9-L11) defined in its existing content script [that already gets injected into every page](https://github.com/keepassxreboot/keepassxc-browser/blob/a037a0b656201fdf9a467c243fd2b6c6a3f9d8b9/keepassxc-browser/manifest.json#L48-L76), so you can probably just reuse that.

Also, it's probably worth noting that other scripts on the same page as the content script are necessarily same-origin with the domain you're authenticating against, so they can, by design, already get access to things like passwords (by reading the `value` attribute of a password input element), or perform authentication using public key credentials (by calling `navigator.credentials.get` themselves). There's very little that can be done to prevent that, nor is it really the responsibility of WebAuthn to do so. It's only really scripts on _other_ origins you need to worry about, and that can be solved using the `sendMessage` API I mentioned above.

Obviously you don't want scripts on the page being able to read the private key, regardless of their origin. The solution to that I believe is to not send the private key to content scripts in the first place. Instead you basically want to run those private parts of the script in the extension's background page/worker and communicate with the content script via [message passing](https://developer.chrome.com/docs/extensions/mv3/messaging/) (e.g. using the sendMessage API I just described).

**Comment by varjolintu** [comment]
@Ajedi32 Thanks for your comment. I had problems passing messages using `browser.runtime.sendMessage` because `browser` namespace is not defined or available in the injected script. Otherwise it would be an ideal solution.

**Comment by Ajedi32** [comment]
Okay, I see the problem you're having.

## Problem summary

Content scripts [can't interfere with the runtime environment of other scripts on the page](https://developer.chrome.com/docs/extensions/mv2/content_scripts/#isolated_world), which you need to do in this case since you need to override the WebAuthn API. So instead you're declaring the WebAuthn override script as a [web-accessible resource](https://developer.chrome.com/docs/extensions/mv2/manifest/web_accessible_resources/) and using a content script to manually inject that into the page.

The problem is that this injected script doesn't have access to `browser.runtime.sendMessage`; since [only content scripts can access private extension APIs](https://developer.chrome.com/docs/extensions/mv2/content_scripts/#capabilities). You could theoretically fix that by declaring the page as [externally connectable](https://developer.chrome.com/docs/extensions/mv2/manifest/externally_connectable/) to allow it to use `postMessage`, except that since WebAuthn could potentially be used on any site you would have to do that for all pages, and declaring all pages as externally connectable is [explicitly disallowed](https://developer.chrome.com/docs/extensions/mv2/manifest/externally_connectable/#reference) for some reason.

That leaves `window.postMessage` as the next most obvious way of communicating between the injected script and the extension, as explained [here](https://developer.chrome.com/docs/extensions/mv2/content_scripts/#host-page-communication). The only problem with that is that other scripts could potentially be listening to those events, and the normal way of preventing that using the `targetOrigin` parameter is, again, [explicitly not allowed in extensions for some reason](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#using_window.postmessage_in_extensions_non-standard).

## Possible solutions

### Solution 1: Use postMessage anyway

The good news is, after giving it some thought I'm not certain those limitations on `targetOrigin` are actually a deal breaker. Obviously leaking the contents of WebAuthn API calls to third-party origins would be _bad_, but I'm not sure that's actually possible as long as you're only calling `postMessage` on your _own_ window. Indeed, the MDN docs [warn not to use a target origin of `*` when you use `postMessage` to send data to other windows](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#security_concerns), but do not give that same warning for sending data to your own window. So assuming I'm not overlooking anything, using `postMessage` might still be viable.

One issue is that you wouldn't be able to send responses back from the extension with `postMessage`, since [browsers apparently set the origin of such requests to null](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#using_window.postmessage_in_extensions_non-standard), making them impossible to verify. That's solvable though by sending a communication channel from the host page using the [Channel Messaging API](https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API) and communicating back via that method rather than via `postMessage` (or alternately, by authenticating the `postMessage` response using a shared secret sent in the request message).

Another other issue is that, as you pointed out, this method doesn't protect against other scripts on the same page intercepting messages to the WebAuth API. As I said though, I don't think that's actually a security issue, since those scripts _should_ all be same-origin with the host page (unless I'm overlooking something) and so could already intercept communications with the WebAuthn API by doing exactly what we're doing (overriding the API with their own functions).

### Solution 2: Custom Events

Another alternative solution is hinted at in [the MDN docs about using postMessage in extensions](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#using_window.postmessage_in_extensions_non-standard), where it suggests using "custom events to communicate with content scripts", "with randomly generated event names, if needed, to prevent snooping from the guest page".

See [the Events article on MDN](https://developer.mozilla.org/en-US/docs/Web/Events/Creating_and_triggering_events) for details on how that would work. Basically you'd create a `CustomEvent` with a randomly-generated name, attach event listeners for that event to some common DOM object that both the content script and injected WebAuthn script have access to, then send messages back and forth by triggering that event on the DOM object. I'm not sure if this is actually more secure than postMessage, but it does seem less likely to cause accidental interference between the extension and host page at least.

**Comment by varjolintu** [comment]
@Ajedi32 Thanks. I've already looked at both solutions before and I agree with you that there's no perfect way to transfer those messages, but Custom Messages is a slightly better solution.
What's a little concerning is that while WebAuthn is intended to be quite secure it doesn't take that much to make a malicious extension that can potentially listen to authentication messages.

The most optimal solution would be that browsers have an API for developers with a separate permission for extensions. This could allow making software authenticators much easier, but I totally understand why things are kept behind closed doors. Still, there's no permissions needed for overriding `navigator.credentials` when using injected scripts.

**Comment by powerman** [comment]
> authenticating the postMessage response using a shared secret sent in the request message

If this shared secret can be detected by others by listening to all messages (or by detecting secret on own page, in case all pages uses same shared secret), then it's probably not a good idea. But there is a usual private/public key solution to this issue.

**Comment by Ajedi32** [comment]
Malicious extensions are a problem in general; the whole security model of the web is based on the same origin policy and extensions violate that by design. Trying to protect the navigator.credentials API specifically wouldn't really help; there are a million ways to steal sensitive information (cookies, passwords, user data, etc) off a site once you have XSS capability there. Chrome has been trying to combat that by discouraging script injection when possible, but for a lot of use-cases (including KeePassXC's) it's an unfortunate necessity.

Ideally yeah, browsers would have dedicated extension APIs for implementing WebAuthn providers and password managers so we wouldn't have to inject scripts at all, at least for this use-case. We're not there yet, unfortunately.

In the mean time I think a solution based on either `postMessage` or Custom Events could get the job done, though I recognize it's a bit more awkward to implement than a `sendMessage`-based solution would be, if that were allowed.

**Comment by Ajedi32** [comment]
> > authenticating the postMessage response using a shared secret sent in the request message
>
> If this shared secret can be detected by others by listening to all messages (or by detecting secret on own page, in case all pages uses same shared secret), then it's probably not a good idea. 

As I explained, I don't think we need to worry about cross-origin scripts being able to listen to messages as long as we're only posting to our own window. The shared secret is just to authenticate that the response came from a script that actually saw the original request, since the [`source` parameter which would normally be used for that purpose apparently gets stripped from the dispatched message when it's sent by an extension](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#using_window.postmessage_in_extensions_non-standard).

...though actually now that I look at it, you might not need the `source` parameter; the [`origin` parameter](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#the_dispatched_event) might be enough to authenticate the response message if you're sending it from a content script, and AFAIK that doesn't get stripped.

**Comment by dancojocaru2000** [comment]
> Ideally yeah, browsers would have dedicated extension APIs for implementing WebAuthn providers and password managers so we wouldn't have to inject scripts at all, at least for this use-case. We're not there yet, unfortunately.

Maybe support for this will arrive following Apple's announcement of Passkeys (which, afaik, is fancy Apple speak for software WebAuthn) and Google's announcement that they will support it?

**Comment by varjolintu** [comment]
> Maybe support for this will arrive following Apple's announcement of Passkeys (which, afaik, is fancy Apple speak for software WebAuthn) and Google's announcement that they will support it?

That's one possibility. I remember I saw a mention about support for software authenticators.

**Comment by VictorVow** [comment]
[Support for WebAuthn is now in Android and Chrome](https://android-developers.googleblog.com/2022/10/bringing-passkeys-to-android-and-chrome.html). Hopefully support in KeePassXC will come for this soon so not too many users migrate.

**Comment by varjolintu** [comment]
> [Support for WebAuthn is now in Android and Chrome](https://android-developers.googleblog.com/2022/10/bringing-passkeys-to-android-and-chrome.html). Hopefully support in KeePassXC will come for this soon so not too many users migrate.

A quick glance at this shows that the API for Passkeys with Chrome is restricted to its own password manager. Which is a bit disappointing.

**Comment by varjolintu** [comment]
I hit a dead end with this feature. I got it to a point where I can actually intercept the registration attempt, direct that to KeePassXC, build a response and send it back to the browser. Then the problems started because [PublicKeyCredential](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential) interface is expected as a response for the registration process, and using that interface is forbidden. The only way that could be used is to rewrite all necessary from the interface to a new class.. which is not a preferred solution.

So far it seems the best solution for this would be to use a OS level software authenticator.

**Comment by varjolintu** [comment]
Just wanted to share that I'm making some progress with this, but unable to say if/when it's ready for even testing.

**Comment by rkjnsn** [comment]
It may now be possible to avoid having to inject a script at all. Chrome (and thus presumably Edge?) now has a (currently mostly undocumented, I believe, unfortunately) webAuthenticationProxy extension API that can theoretically be used to intercept WebAuthn requests, and on Firefox the extension could use `wrappedJSObject` and `exportFunction()`/`cloneInto()` to hook the WebAuthn API directly from the content script.

**Comment by rugk** [comment]
Hmm [could not find any support data or doc for that on Chromium](https://chromestatus.com/features#web%20authentication) BTW. Also no WebExtension API doc for that webAuthenticationProxy API.

**Comment by rkjnsn** [comment]
Yes, it unfortunately doesn't appear to be documented at all. The implementation is here: https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/extensions/api/web_authentication_proxy/web_authentication_proxy_api.h

An API test extension exercising the API is here: https://source.chromium.org/chromium/chromium/src/+/main:chrome/test/data/extensions/api_test/web_authentication_proxy/main/test.js

The gist seems to be that the extension adds listeners to the various events `onCreateRequest`, `onGetRequest`, `onIsUvpaaRequest`, and `onRequestCanceled`, and then calls `attach`. In the event handlers (except for `onRequestCanceled`), carry out the requested action and call the respective complete function with the result (e.g., `completeCreateRequest`, `completeIsUvpaaRequest`, et cetera).  There's also an `onRemoteSessionStateChange`, which appears to provide a way for a native application to kick the extension when needed. This could hypothetically be used to have the extension detach when the database is locked or KeepassXC exits, and reattach when the database is unlocked.

**Comment by varjolintu** [comment]
@rkjnsn As far as I know that API cannot be accessed via browser.

**Comment by rkjnsn** [comment]
I haven't tried to implement it, but I can definitely see the API if I inspect a dummy extension with the appropriate permission: 
![screenshot](https://user-images.githubusercontent.com/7364197/197946496-22169959-2544-4276-a6a0-a6497bd00510.png)

**Comment by varjolintu** [comment]
@rkjnsn Interesting. Let's hope that API is something we can actually use instead of the dirty inject script version. And it's possible to release an extension that uses that specific permission to access the API.

**Comment by varjolintu** [comment]
At this point I have a KeePassXC version that can build (and decode) a proper response for the WebAuthn register request. So the hardest part is already done.

**Comment by varjolintu** [comment]
1/2 of the steps done using webauthn.io as a test site :)

<img width="700" alt="webauthn_register" src="https://user-images.githubusercontent.com/24570482/198693225-420e251c-ccef-45a7-9af9-15aeba45c696.png">

**Comment by Ajedi32** [comment]
I just wanted to say thanks for the great work on this so far!

Now that browsers are starting to implement synced WebAuthn credentials themselves under the banner of passkeys, this feature is about to become way more useful as mainstream sites start using passkeys as their primary authentication method.

Furthermore, independent cross-browser implementations like KeePassXC are going to be really important for avoiding issues of vendor lock-in in the short term, since from what I've seen so far most of the browsers implementing passkeys right now don't support any mechanism of transferring the key database to another browser.

**Comment by varjolintu** [comment]
@Ajedi32 The next step here is of course to decide how the key is stored in KeePassXC so it can be easily exported as a file if needed. Importing keys to another database (or some other password manager) would be quite helpful.

EDIT: The proper format for the key is of course https://en.wikipedia.org/wiki/PKCS_8.

**Comment by varjolintu** [comment]
<img width="476" alt="webauthn_get" src="https://user-images.githubusercontent.com/24570482/199530061-897f322a-c1dd-4aab-841b-7a132b1e6db6.png">

One more step achieved.

**Comment by varjolintu** [comment]
One thing that will make this feature a bit problematic is the signature count which is updated every time a successful authentication is made. This means the database is updated every time the keys are used, which can cause problems to those users who are syncing the database manually for multiple devices. If the count added to the response is too low, authentication with the web site fails.

So for now I'm making the signature count as an attribute which can be easily edited if needed. Not the optimal solution, but it could help.

**Comment by zroug** [comment]
I think signature counters are optional and not supporting them has gotten more common since the introduction of passkeys, because they face the same problem. https://github.com/w3c/webauthn/issues/1734 talks a bit about this. Maybe just leaving the counter at zero is an option? However, my knowledge on this is very limited, so please excuse me if I'm mistaken on this. I just wanted to mention it, but I don't know if it is viable.

**Comment by droidmonkey** [comment]
I like the start at 0 stay at 0 approach.

**Comment by varjolintu** [comment]
The site I'm using to test this feature has a mandatory signature counter. The request from the server doesn't include any flag or variable that indicates if the counter is used or not, so staying at 0 will fail the authentication. There must be an option to increase it.

**Comment by droidmonkey** [comment]
Seems to be a bug in the test then because according to the github discussion it is sn optional feature.

**Comment by Ajedi32** [comment]
Here's the relevant portion of the latest WebAuthn draft:

[Web Authentication Editor’s Draft, 10 October 2022, Section 6.1.1. Signature Counter Considerations](https://w3c.github.io/webauthn/#sctn-sign-counter):

> Authenticators that do not implement a [signature counter](https://w3c.github.io/webauthn/#signature-counter) leave the [signCount](https://w3c.github.io/webauthn/#authdata-signcount) in the [authenticator data](https://w3c.github.io/webauthn/#authenticator-data) constant at zero.

Even [the current published version of the standard](https://www.w3.org/TR/webauthn-3/#sctn-sign-counter) suggests relying parties shouldn't complain if the counter was registered as zero and stays that way:

> the [Relying Party](https://w3c.github.io/webauthn/#relying-party) compares the stored [signature counter](https://w3c.github.io/webauthn/#signature-counter) value with the new [signCount](https://w3c.github.io/webauthn/#authdata-signcount) value returned in the assertion’s [authenticator data](https://w3c.github.io/webauthn/#authenticator-data). **_If either is non-zero_**, and the new [signCount](https://w3c.github.io/webauthn/#authdata-signcount) value is less than or equal to the stored value, a cloned authenticator may exist, or the authenticator may be malfunctioning.

(Emphasis mine.)

Does the site you're testing with still fail even in that case?

**Comment by varjolintu** [comment]
Thanks for the info! If the counter is set to `1` at the register phase, the page expects the incrementation. By setting it to zero everything works like a charm.

**Comment by varjolintu** [comment]
Just a small update: I plan to put PR's public during this month so people can start testing this. The feature works pretty nicely now, but the following things are still missing:
- Support for extensions on register (some sites with Passkeys likes to use these)
- Support for Firefox (some API differences in the browser)
- Support for a new confirmation dialog (with a normal confirmation, or Windows Hello / TouchID with quick unlock enabled + handling multiple credentials)

I have been these sites for testing: https://webauthn.io/, https://webauthn.lubu.ch/_test/client.html and https://demo.yubico.com/webauthn-technical/registration. All of them work flawlessly.

**Comment by varjolintu** [comment]
Here's the first draft. Feel free to test it:
https://github.com/keepassxreboot/keepassxc/pull/8825
https://github.com/keepassxreboot/keepassxc-browser/pull/1786

**Comment by nagromc** [comment]
I’d like to provide feedback, but I use the stable build ATM (installation via Microsoft Store or Scoop, depending on the machine). Am I supposed to build the branch on my own, or are there any nightly I can give a try?

**Comment by varjolintu** [comment]
@nagromc For now you'll have to build the branch by yourself. Of course I can provide the binaries too if wanted, but then you'll have to make a leap of trust :)

**Comment by mkatsevVR** [comment]
Curious what the status of this is. Are you still working on the implementation or is it done and mainly waiting on the maintainers to merge the PR?

**Comment by varjolintu** [comment]
@mkatsevVR It's not 100% yet. But there are some things that must be resolved before this can be even merged. For example our CI server is using an old Qt version so we would need a separate build just for this feature that includes a newer Qt version. And it would be nice to confirm this could work with multiple sites that provide support for Passkeys.

Still, even if this feature would be approved, it would be in KeePassXC 2.8.0 (the earliest possible version) and the release data for that is still probably far in the future. Maybe this year? Cannot say for sure.

**Comment by JamieOrlando-Engooden** [comment]
@varjolintu Anything we can do as a community to help with those road blocks. Having keepassxc support passkeys will be a huge win for everyone.

**Comment by varjolintu** [comment]
@JamieOrlando-Engooden At least it would be helpful to test the feature with sites that support WebAuthn/Passkeys. I've tried some outside the four sites in the PR but those are often working with a pure luck. Some sites just doesn't work, and there's no valid reply from the server or any of the site's messages. Haven't checked the official site list for a while though.

**Comment by haldi4803** [comment]
> @nagromc For now you'll have to build the branch by yourself. Of course I can provide the binaries too if wanted, but then you'll have to make a leap of trust :)

Would appreciate that :)
For simple testing of passkeys in a separate database you certainly seem trustworthy enough.

**Comment by varjolintu** [comment]
> > @nagromc For now you'll have to build the branch by yourself. Of course I can provide the binaries too if wanted, but then you'll have to make a leap of trust :)
> 
> Would appreciate that :) For simple testing of passkeys in a separate database you certainly seem trustworthy enough.

See: https://github.com/varjolintu/keepassxc/releases/tag/2.8.0-webauthn

**Comment by haldi4803** [comment]
Successfully tested that with CouldFlare's Keypass 2FA :)

BestBuy says my device or browser (chromium) doesn't support Fingerpring/Face Sign-In for Passkeys.

Here is a list of other Homepages that support Passkeys: https://passkeys.directory/

formextractorai.com/login seems to be somewhat bugged, as it DOES register the passkey but the hompage doesn't regocnize the passkey being saved and doesn't forward/continue. So you're just stuck there.
Mangadex.org which uses the keycloak service brings the "Error: Credential is excluded" message
Microsoft.com didn't dedect KeepassXC as Passkey forced phone or ubikey instead.
hyatt.com didnt do anything when pressing "create passkey" button...

**Comment by JohnLGalt** [comment]
> > > @nagromc For now you'll have to build the branch by yourself. Of course I can provide the binaries too if wanted, but then you'll have to make a leap of trust :)
> > 
> > 
> > Would appreciate that :) For simple testing of passkeys in a separate database you certainly seem trustworthy enough.
> 
> See: https://github.com/varjolintu/keepassxc/releases/tag/2.8.0-webauthn

I use KeePassXC  via portable mode, and I'm willing to give a shot at building this specialized version for testing Passkeys.

I'm just not all that up to snuff on what I need to build, and I'd prefer **not** to have to install it (even via MSI).

Any pointers, @varjolintu ? I'm looking for build instructions specifically to build a portable version (of KPXC, since I'm already using the development snapshot releases as my daily driver). I use Firefox as my primary browser, both stable and Nightly versions, and I'm well versed in testing addons in distinct profiles to prevent issues with my daily use profiles.

**Comment by varjolintu** [comment]
@JohnLGalt I can share the portable version as well tomorrow.

**Comment by JohnLGalt** [comment]
Thanks! That would be helpful!

However, if this is going to be ongoing in terms of development as we test stuff, I should probably learn to build it, eh? Then you won't have to build a portable every time....

If you don't mind building, cool. But I'm more than willing to actively test including building.

**Comment by varjolintu** [comment]
@JohnLGalt I'm not planning to do new builds frequently. Only if something bigger is updated. I have already received some reports on sites that work and are not working, so there's some debugging to do.

**Comment by varjolintu** [comment]
With the latest additions to the repos the Passkeys feature now works with GitHub as well, with both 2FA and as a login Passkey, depending which one user chooses to use.

**Comment by pabs3** [comment]
I have a slightly but hopefully not too offtopic question; currently WebAuthn cannot be used when JavaScript is disabled, have the KeePassXC folks thought about how to support WebAuthn in that scenario, or with websites that don't want to ship JavaScript?

https://github.com/w3c/webauthn/issues/1255

**Comment by varjolintu** [comment]
> I have a slightly but hopefully not too offtopic question; currently WebAuthn cannot be used when JavaScript is disabled, have the KeePassXC folks thought about how to support WebAuthn in that scenario, or with websites that don't want to ship JavaScript?
> 
> [w3c/webauthn#1255](https://github.com/w3c/webauthn/issues/1255)

I don't think that's possible, unless browsers itself add a feature where it's possible to pass WebAuthn calls directly outside the browser (e.g. using native messaging). So far our only solution for this is to use JavaScript for capturing the necessary API calls.

**Comment by varjolintu** [comment]
And with the latest changes Google's Passkey support also works with KeePassXC. And of course when trying the feature with Firefox, Google complains the browser is not compatible ;) Didn't try it yet with a different user agent though.

EDIT: Microsoft's site also works.

**Comment by FlyveHest** [comment]
Google now defaults to using passkeys when signing in.

https://blog.google/technology/safety-security/passkeys-default-google-accounts/

**Comment by user858753257** [comment]
When you publish the Passkeys support . Will be Linux also supported ?

**Comment by varjolintu** [comment]
> When you publish the Passkeys support . Will be Linux also supported ?

Linux is of course supported, with all browsers using the browser extension.

**Comment by user858753257** [comment]
> > When you publish the Passkeys support . Will be Linux also supported ?
> 
> 
> 
> Linux is of course supported, with all browsers using the browser extension.

Perfect . Until now that's a missing point for me on Linux systems 🙏

When I read it right the kernel itself doesent have support

**Comment by Midou36O** [comment]
![image](https://github.com/keepassxreboot/keepassxc/assets/45198486/925a32a3-7d32-48c4-9684-199e58cd5782)

Is it backwards compatible with hardware security key APIs? Some websites like the yubikey website reports successful registration and login but gitea reports a successful registration but failure at login.

---

**Update**: Nevermind, it's just a problem on my end, using passkeys on a different domain from the original one that was registered from is not permitted. Marking this as resolved.

**Comment by varjolintu** [comment]
Anyone who missed the messages in the PR, you can download the latest snapshot build (2023-10-26 or later) from https://snapshot.keepassxc.org/ and test the Passkey feature.

(Use a test database and/or have some decent backups, just in case)

**Comment by opotonniee** [comment]
Congrats for this new feature, very useful!

A few comments after testing the nov-5 snapshot:

- After upgrading from 2.7.6, the execution failed with several error popups complaining about missing Qt5 DLLs. I uninstalled and reinstalled and then it worked fine.
- Although it is supposed to be opaque, it is confusing that you call KPEX_PASSKEY_GENERATED_**USER_ID** the credential Id. The User ID is actually the user handle
- It is a pity it does not support passkey [autofill](https://passkeys.dev/docs/use-cases/bootstrapping/#authenticating-the-user), is this in the plans? 
- Why does #8825 state that it does not support resident keys? Do you mean it always creates resident key, whatever the request attributes?

**Comment by varjolintu** [comment]
>   Although it is supposed to be opaque, it is confusing that you call KPEX_PASSKEY_GENERATED_**USER_ID** the credential Id. The User ID is actually the user handle

Not sure what you mean here. Could you clarify this a bit?

>  It is a pity it does not support passkey [autofill](https://passkeys.dev/docs/use-cases/bootstrapping/#authenticating-the-user), is this in the plans?

Passkey should be supported as well. See https://github.com/keepassxreboot/keepassxc/pull/9987 for some small related changes.

>  Why does [Add basic support for WebAuthn (Passkeys) #8825](https://github.com/keepassxreboot/keepassxc/pull/8825) state that it does not support resident keys? Do you mean it always creates resident key, whatever the request attributes?

We are always storing username, user ID etc. but not handling situations where `requireResidentKey` is set.

**Comment by opotonniee** [comment]
> > Although it is supposed to be opaque, it is confusing that you call KPEX_PASSKEY_GENERATED_**USER_ID** the credential Id. The User ID is actually the user handle
> 
> Not sure what you mean here. Could you clarify this a bit?

I mean the advanced attribute named KPEX_PASSKEY_GENERATED_USER_ID is actually storing the credential id, not the user id. This attribute name is misleading.

> > It is a pity it does not support passkey [autofill](https://passkeys.dev/docs/use-cases/bootstrapping/#authenticating-the-user), is this in the plans?
> 
> Passkey should be supported as well. See #9987 for some small related changes.

This PR is about importing passkeys, I don't see how it addresses autofill, which is about prompting the user to select an already registered passkeys when an authentication request does not contain the user name.

> 
> > Why does [Add basic support for WebAuthn (Passkeys) #8825](https://github.com/keepassxreboot/keepassxc/pull/8825) state that it does not support resident keys? Do you mean it always creates resident key, whatever the request attributes?
> 
> We are always storing username, user ID etc. but not handling situations where `requireResidentKey` is set.

ok, clear.

**Comment by varjolintu** [comment]
> I mean the advanced attribute named KPEX_PASSKEY_GENERATED_USER_ID is actually storing the credential id, not the user id. This attribute name is misleading.

Ah, you are meaning this one? https://w3c.github.io/webauthn/#ref-for-dom-credential-id

> This PR is about importing passkeys, I don't see how it addresses autofill, which is about prompting the user to select an already registered passkeys when an authentication request does not contain the user name.

It includes a change where Relying Party is checked with the normal URL, exposing passkeys to the normal autofill. It's also filling the username defined in Passkey instead of the entry's username. The current snapshot version does not do this, except if the Passkey is stored to a new entry. If an authentication request doesn't contain any allowed credentials, all entries with user handle are returned to the autofill.

**Comment by taelfrinn** [comment]
> We are always storing username, user ID etc. but not handling situations where `requireResidentKey` is set.

my understanding of non-resident keys is that they are not stored. they are generated on demand from information stored in the id which is why non-resident keys tend to have huge ids. (the credentialid contains some seed data, which can be combined with the RP and username to resynthesize the key. it might also have some kind of mac to prove it hasnt been tampered with). So the website has to present a list of "allowCredentials" in order for the fido device to re-create keys from.

While resident keys rely on local storage, and the RP does not provide any kind of keyids or userid to the user, and lets the client choose from any stored keys for the RP. Given that this implementation stores keys, full support for resident keys AKA passkeys should be possible. Full resident keys are what constitutes passkeys, while non-resident keys are just a 2fa method - their design being typically used after password login completes, as MFA to it.

Also, by extension, it should also be possible to generate real "non-resident" keys in which nothing is stored locally.  Of course, given that the keepassvault does not have the extreme space limitations of hardware tokens, this is probably not worth any effort.

-=-=-

Separately, one prerequisite here that is not mentioned is that the master vault secret should be at least as secure as the keys it contains. (i.e. not chosen by the user, at least having an entropy of 128 bits) Given that passkeys will replace passwords, a weak and easily guessed user-chosen master vault passphrase should probably not be an easy option for users, and instead we should push/insist on/or default to a "correct horse battery staple"-like spec, such as bip39, for master vaults. (which have the property that a user cannot just pick them manually, but must synth them from entropy)

This is important here because passkeys generally dont need 2fa. Unlike a traditional keepass vault which can be weak user-chosen passwords protected by a weak user-chosen vault password, then "backed" by SMS or OTP 2fa, a database of passkeys will usually mean direct and full access. 

External USB keys deal with this by being a separate tamper resistant device. A software alternative has to have at least a crypto-strong on-disk vault, imo, due to being exposed to all other software the user may run.

IMO, a software version of webauthn, whether resident passkeys or non-resident 2FA methods, should at a minimum require 128 bits of entropy for the vault.

**Comment by phoerious** [comment]
> Separately, one prerequisite here that is not mentioned is that the master vault secret should be at least as secure as the keys it contains. (i.e. not chosen by the user, at least having an entropy of 128 bits) Given that passkeys will replace passwords, a weak and easily guessed user-chosen master vault passphrase should probably not be an easy option for users, and instead we should push/insist on/or default to a "correct horse battery staple"-like spec,

That's incorrect. Your master password does not need to be as secure as the PassKey and enforcing such a thing will alleviate any benefit you get from it. KeePassXC transforms the Key with Argon2 for exactly that reason. This doesn't make up for very weak passwords, but you also don't need a super-random 128-bit string for reasonable security. Your password database is also offline on your machine, so an attacker needs access to the device. In any case, even a weak password is more secure than simply storing it in your browser without additional protection.

**Comment by ccoenen** [comment]
I fully agree with @phoerious. If you do PassKey with a Security Token like a YubiKey, you also don't input a 128+ bit sequence via that one button. (and even if you unlocked a PassKey store on your phone with a fingerprint, there is [(much) less than 128 bits of information in your fingerprint](http://www.cedar.buffalo.edu/~govind/CSE717/papers/CancelableBiometrics.pdf)!)

**Comment by ypurpl** [comment]
Just to note, Bitwarden has passkey support slowly rolling out. This should be very much doable.

**Comment by MrConorAE** [comment]
Hey there, just started using KeePass from 1Password and this looks super interesting. Is there any timeline on getting this to stable, or is it going to be in snapshots only for now?

**Comment by droidmonkey** [comment]
No hard timeline right now, but I would think shortly after the new year.

**Comment by andycandy-de** [comment]
I am also waiting for this great feature 🤣
Thank you for your great work ♥️

**Comment by gwillen** [comment]
I'm trying to test this out in the snapshot. What I've found so far:
- On Linux, it works great with webauthn.io (in both Firefox and Chrome), but Google refuses to generate a passkey ("can't be created on this device".) I don't know whether that's expected, or whether other password manager extensions are able to work around it.
- On macOS, I'm unable to run the snapshot on my machine, because it seems to have a minimum version of macOS 13. I'm not sure whether that's intended, or an artifact of the snapshot build process being different from the release build process.

EDIT: However, your build process is unusually pleasant and smooth for a mac app, so I just built from source, and everything seems to work great! The UX is a little unintuitive when logging in with a passkey on google.com, but that seems unsurprising at this early stage.

**Comment by varjolintu** [comment]
> I'm trying to test this out in the snapshot. What I've found so far:
> 
>     * On Linux, it works great with webauthn.io (in both Firefox and Chrome), but Google refuses to generate a passkey ("can't be created on this device".) I don't know whether that's expected, or whether other password manager extensions are able to work around it.
> 
>     * On macOS, I'm unable to run the snapshot on my machine, because it seems to have a minimum version of macOS 13. I'm not sure whether that's intended, or an artifact of the snapshot build process being different from the release build process.
> 
> 
> EDIT: However, your build process is unusually pleasant and smooth for a mac app, so I just built from source, and everything seems to work great! The UX is a little unintuitive when logging in with a passkey on google.com, but that seems unsurprising at this early stage.

With Google, do you mean with Firefox? I was unable to create a Passkey for Google with Firefox. It's probably a user-agent related restriction. Any Chromium-based browser works just fine.

About the UX, maybe we shouldn't ask a separate confirmation for Passkey entries because those are confirmed anyway in a separate dialog.

**Comment by kwin** [comment]
Firefox has a bug with Google passkeys: https://bugzilla.mozilla.org/show_bug.cgi?id=1865625.

**Comment by gwillen** [comment]
It looks like that Firefox bug is specifically for macOS -- on my macOS machine, I only tried Edge (and it worked). My failures were both on Linux, where it failed in both Chrome and Firefox. (My vague understanding is that this is expected on Linux.)

For the UX issues: creating a Passkey was fine as I recall, but when logging in with a Passkey I get the following sequence of events (some of which are probably Google's fault):
- (Assuming the database is already unlocked:)
- I get a "Browser Access Request" and I select a passkey and hit "Allow Selected".
- This has no visible effect on the page; the email field is still blank. (Clicking the keepassxc icon in the field then populates it. This interaction seems like it shouldn't be necessary given the previous one.)
- I click "next", and google gives me a password field.
  - This ... seems like google's fault? I don't know why it's not automatically asking for passkeys, since the account is set up for passkeys. But it could also be that it's looking for something in the environment that it's not finding?
  - There is a keepassx icon in the password box, but it's a red herring; I don't have a password saved for this account so it doesn't do anything.
- I click "Try another way" -> "Use your passkey".
- Then I get the authentication prompt and everything works from there.

(This is all when doing a first-time login on a new device, simulated using incognito mode. If I log out and go to log back in, everything is seamless -- it gives a list of logged-out accounts, I click the one I want, it gives a dialog mumbling about using a passkey, I click next, the authentication happens.)

**Comment by foss-** [comment]
> Firefox has a bug with Google passkeys: https://bugzilla.mozilla.org/show_bug.cgi?id=1865625.

Note that this is fixed already and fix is available in v122 (currently nightly branch). So you could install Firefox nighty to test on macOS.

**Comment by sovamind** [comment]
Passkey support might be broken on Windows 11 22H2.  Several of us are reporting that it no longer works properly and I noticed it after installing the 22H2 update.  We have another issue thread tracking it here: https://github.com/keepassxreboot/keepassxc/issues/11160

**Comment by m-a-v** [comment]
No issues on my side. Just tested with the newest version 2.7.9. Here you refer to 22H2, but in the bug issue 24H2 is referred which I think is the version you wanted to mention.
